### PR TITLE
Adding ArrayDerivative class as subclass of Derivative, will handle d…

### PR DESCRIPTION
**Pull request by Upabjojr**
Adding ArrayDerivative class as subclass of Derivative, will handle derivatives involving non scalar expressions.

Getting rid of the visitor pattern in order to perform double-dispatch and decide how to handle derivatives.

This PR does not change the logic of SymPy, it's just code clean-up.

#### Release Notes

* tensor
  * Adding ArrayDerivative class as subclass of Derivative. This new handles derivatives involving non-scalar expressions.

### CODE REVIEW

File: sympy/tensor/array/array_derivatives.py [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+
+
+class ArrayDerivative(Derivative):
+
+    is_scalar = False
+
+    def __new__(cls, expr, *variables, **kwargs):
+        obj = super(ArrayDerivative, cls).__new__(cls, expr, *variables, **kwargs)
+        if isinstance(obj, ArrayDerivative):
+            obj._shape = obj._get_shape()
```
Comment (oscarbenjamin) [inline sympy/tensor/array/array_derivatives.py:17]: There is a `cached_property` decorator in ode.py for this sort of thing:
https://github.com/sympy/sympy/blob/b6c328e2f057ba81513d19a1451ace358ecc8233/sympy/solvers/ode/single.py#L36-L45

Perhaps that could be moved to `sympy/core/decorators.py` and then reused for `shape` here.

Comment (Upabjojr) [reply sympy/tensor/array/array_derivatives.py:17]: Well, concatenating a few tuples may be computationally less expensive than caching a property.

Comment (oscarbenjamin) [reply sympy/tensor/array/array_derivatives.py:17]: I will be surprised if you can measure a difference in any real calculation.

The class here is already using a cached property. The shape property returns this `_shape` attribute which is precomputed in the constructor. The `cached_property` decorator exists to make this pattern as simple as applying a decorator to the property method and also has the advantage of not computing the result if it is never accessed. That's a big performance improvement in some other situations although likely insignificant in this case.

Comment (Upabjojr) [reply sympy/tensor/array/array_derivatives.py:17]: OK, maybe in a separate PR.

**Comment by Upabjojr** [comment]
I'm going to merge this PR if nobody objects.

**Comment by oscarbenjamin** [comment]
What exactly is the relationship between sympy's array and tensor types?

I would have expected matrix derivatives to give tensors rather than arrays since generalisations of matrix differentiation are usually based on considering a matrix to be a rank-2 tensor.

**Comment by Upabjojr** [comment]
To sum up our current types:

- classes inheriting `MatrixCommon`: explicit-valued matrices (all elements of the matrix are expressed),
- classes inheriting `MatrixExpr` but not `MatrixCommon`: implicit-valued matrices (i.e. the matrix is just a symbol, no elements are shown),
- `MatrixElement` class: I'd call it implicit-valued (as elements are not shown), explicit-indexed (as indices are shown, unlike `MatrixExpr`),
- classes inheriting `NDimArray`: like explicit matrices, but with N-axes (matrices always have 2 axes),
- `IndexedBase` and `Indexed` objects: generic indexed objects, implicit-valued explicit-index,
- classes in `sympy.codegen.array_utils`: implicit-valued implicit-indexed arrays, still experimental.
- `sympy.tensor.tensor`: abstract index notation for tensorial expressions. This is reserved for applications in theoretical physics.


> I would have expected matrix derivatives to give tensors rather than arrays since generalisations of matrix differentiation are usually based on considering a matrix to be a rank-2 tensor.

Define what you mean by tensor. N-dimensional arrays are a generalization of matrices to more (or less) than 2 axes.

With _tensor_ we should stick to the mathematical definition in SymPy. The term is being conflated with N-dim arrays in computer science (e.g. Google's Tensorflow library).

Technically a tensor is defined on a manifold and has covariant and contravariant components. When expressed in some coordinate systems it's an N-dimensional array, but covariant and contravariant components will transform differently under coordinate system changes.

**Comment by Upabjojr** [comment]
> I would have expected matrix derivatives to give tensors rather than arrays since generalisations of matrix differentiation are usually based on considering a matrix to be a rank-2 tensor.

In some cases the matrix derivative is still a matrix. For example, the trace of a matrix derived by a matrix is a matrix. If a matrix has shape `(k, 1)` and is derived by a matrix of shape `(k, 1)`, you may get the shape `(k, 1, k, 1) ==> (k, k)` matrix by dropping trivial, therefore expressing the result as a matrix.

Some combinations of 3-ranked and 4-ranked N-dim arrays may still be represented as matrices with the help of the `TensorProduct` object.

**Comment by oscarbenjamin** [comment]
Okay so I guess that a Tensor is a multilinear map that can be represented as an array with respect to particular bases in the same way that a linear map can be represented as a matrix which is also a 2D array.

It seems like this construction treats a matrix as a special case of an array. Is there an algebra for arrays that can generalise matrix multiplication?

> If a matrix has shape `(k, 1)` and is derived by a matrix of shape `(k, 1)`, you may get the shape `(k, 1, k, 1) ==> (k, k)` matrix by dropping trivial, therefore expressing the result as a matrix.

I'm not sure I follow but identifying shape `(k, 1, k, 1)` as `(k, k)` sounds to me like the kind of convenience that leads to inconsistency later.

**Comment by Upabjojr** [comment]
> It seems like this construction treats a matrix as a special case of an array. Is there an algebra for arrays that can generalise matrix multiplication?

We had a discussion some time ago and decided to only allow the operator `*` (i.e. `__mul__`) between scalars and arrays. `array * array` will raise an exception. I would prefer to keep it like this, as it's not really clear how to generalize the matrix products (for example, should you contract the second axis of the first array with the first axis of the second array, or the last axis of the first array with the first axis of the second array? If the axes are 2, like in matrices, it's not an ambiguous decision).

**Comment by Upabjojr** [comment]
> I'm not sure I follow but identifying shape `(k, 1, k, 1)` as `(k, k)` sounds to me like the kind of convenience that leads to inconsistency later.

That only works for matrix expressions (we are trying to keep matrix derivatives closed inside the `sympy.matrices.expressions` module).

**Comment by Upabjojr** [comment]
Anyway, we have `tensorproduct` and `tensorcontraction`, which are able to perform most of tensor algebra operations if properly combined.

File: sympy/tensor/array/array_derivatives.py [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+    @property
+    def shape(self):
+        return self._shape
+
+    @classmethod
+    def _get_zero_with_shape_like(cls, expr):
+        if isinstance(expr, (MatrixCommon, NDimArray)):
+            return expr.zeros(*expr.shape)
+        elif isinstance(expr, MatrixExpr):
+            return ZeroMatrix(*expr.shape)
```
Comment (eric-wieser) [inline sympy/tensor/array/array_derivatives.py:39]:
```suggestion
    def _get_zero_with_shape_like(cls, expr):
        if isinstance(expr, (MatrixCommon, NDimArray)):
            return expr.zeros(*expr.shape)
        elif isinstance(expr, MatrixExpr):
            return ZeroMatrix(*expr.shape)
        else:
            raise RuntimeError("...")
```
What if neither condition is true? returning `None` seems unlikely to be useful

**Comment by oscarbenjamin** [comment]
@Upabjojr it looks like this PR has caused a slowdown in differentiation (https://github.com/sympy/sympy/issues/20234#issuecomment-706758408)

**Comment by Upabjojr** [comment]
The `.diff( )` method needs to support both scalars and matrices. The code was quite messy, so there was some need to clean up. If people are sure they only need scalar differentiation, maybe using `Derivative( ... )` could avoid passing through the dispatcher.

**Comment by oscarbenjamin** [comment]
There has to be a better way than this. The `sympy.core.function` module should not have so many references to classes like `NDimArray` from outside of core.

**Comment by Upabjojr** [comment]
> There has to be a better way than this. The `sympy.core.function` module should not have so many references to classes like `NDimArray` from outside of core.

Yes, definitely.

**Comment by JSS95** [comment]
> To sum up our current types:

I think this should be documented on somewhere visible.

**Pull request merged**